rust/connlib/clients/shared/src/lib.rs

```diff
@@ -60,20 +63,34 @@ impl Session {
 
     /// Attempts to reconnect a [`Session`].
     ///
-    /// This can and should be called by client applications on any network state changes.
-    /// It is a signal to connlib to:
+    /// Reconnecting a session will:
+    ///
+    /// - Close and re-open a connection to the portal.
+    /// - Refresh all allocations
+    /// - Replace the currently used [`Sockets`] with the provided one
     ///
-    /// - validate all currently used network paths to relays and peers
-    /// - ensure we are connected to the portal
+    /// # Implementation note
     ///
-    /// Reconnect is non-destructive and can be called several times in a row.
+    /// The reason we replace [`Sockets`] are:
     ///
-    /// In case of destructive network state changes, i.e. the user switched from wifi to cellular,
-    /// reconnect allows connlib to re-establish connections faster because we don't have to wait for timeouts first.
-    pub fn reconnect(&self) {
-        let _ = self.channel.send(Command::Reconnect);
+    /// 1. On MacOS, as socket bound to the unspecified IP cannot send to interfaces attached after the socket has been created.
+    /// 2. Switching between networks changes the 3-tuple of the client.
+    ///    The TURN protocol identifies a client's allocation based on the 3-tuple.
+    ///    Consequently, an allocation is invalid after switching networks and we clear the state.
+    ///    Changing the IP would be enough for that.
+    ///    However, if the user would now change _back_ to the previous network,
+    ///    the TURN server would recognise the old allocation but the client already lost all its state associated with it.
+    ///    To avoid race-conditions like this, we initialize a new [`Sockets`] instance which allocates a new port.
+    pub fn reconnect(&self, sockets: Sockets) {
+        let _ = self.channel.send(Command::Reconnect(sockets));
     }
 
+    /// Sets a new set of upstream DNS servers for this [`Session`].
+    ///
+    /// Changing the DNS servers clears all cached DNS requests which may be disruptive to the UX.
+    /// Clients should only call this when relevant.
+    ///
+    /// The implementation is idempotent; calling it with the same set of servers is safe.
```

Comment:
If we're going to test this from the client-side we don't need this here, in which case we don't need to keep track of the system's resolver in connlib.

Reply:
Do you think it is worth getting rid of that state tracking? I am a bit torn. It seems easy to make it idempotent which means a little less code on the clients?

Happy to go either way here.

Reply:
> don't need to keep track of the system's resolver in connlib.

At all? Or just the last few. We'll still have the `set_dns` right?

Reply:
>At all? Or just the last few. We'll still have the set_dns right?

we would still have it, but we only store the `system_resolvers` in connlib to make `set_dns` not react to the same servers

Reply:
> Do you think it is worth getting rid of that state tracking? I am a bit torn. It seems easy to make it idempotent which means a little less code on the clients?

I just think that we shouldn't keep the same logic in two places and since it seems we are going with the route of having this logic in the client I'd just remove it from connlib.